ls1intum/Artemis
# Feature: Import programming exercises

### Checklist

### Motivation and Context


As in [ARTEMIS-471](https://jirabruegge.in.tum.de/browse/ARTEMIS-471?filter=-1) described, we want to support the import of programming exercises from all available (also older) courses. It should be possible to search for any existing exercise and import it into an existing course.

### Description

There are changes both on the client and server side. 
As for the server:

- Added a new endpoint for importing programming exercises, which is available to instructors and admins.
- Mainly modified the `ProgrammingExerciseService` allowing the import of the programming exercise itself, the Bamboo build plans and the Bitbucket repositories. 
- Primitive properties, s.a. the flags for the online code editor, or publishing the build plan just get copied
- Referenced attributes, s.a. the templates or the test cases get cloned and saved in order to generate new entities and identifiers
- Identifiers like the build plan key get generated again based on the target course and new exercise name
- New pageable search functionality allowing users to query all programming exercises and receiving only a page of the result (query by title of the exercise, or the course title; page size and page number are defined by the caller)

As for the client:
- New import button right next to the "Create new programming exercise" button
- New pop-up with a pageable table containing all available programming exercises. 
- The user can search for the title, or the course title of the exercise he wants to import.
- The user can sort the columns of the table
- After selecting an exercise, the programming-exercise-update component is reused for modifying the to be imported exercise and saving it.

About the pageable search: For every new character in the search bar and every time the user sorts the table, there is a call to the backend. This can result in a lot of requests. I.m.o. this should be no problem since the import functionality will only be used in the case of setting up a new course/exercises for that course. Moreover, by doing it this way, we only sent the relevant portion of the exercise table to the user, which might be more important in the future, when there are hundreds of old programming exercises.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Select an existing course
4. Click on "Import new programming exercise"
5. Search for the exercise you want to import and click on "Import"
6. Modify any values for the new exercise
7. Click on "Save"
8. The exercise should be imported and all references, repositories and plans should be copied. (Only the template repos and plans, no student participations) 

### Screenshots


![import_exc](https://user-images.githubusercontent.com/27979674/64736190-609f5000-d4ea-11e9-99d5-1dbdfca6c8cc.gif)
